pipeline/metadata/ip_metadata_chooser.py

```diff
@@ -0,0 +1,110 @@
+"""A module which contains a number of sources of ip metadata and chooses how to add them to ips."""
+import datetime
+import logging
+from typing import Type
+
+from pipeline.metadata import caida_ip_metadata
+from pipeline.metadata import dbip
+from pipeline.metadata import maxmind
+from pipeline.metadata.flatten import Row
+
+
+# pylint: disable=too-many-instance-attributes
+class IpMetadataChooser():
+  """Business logic for selecting IP metadata values from a number of sources"""
+
+  def __init__(self, caida_ip_metadata_class: Type[
```

Comment:
Instead of this class, provide a function that returns a factory function to avoid the partial initialization.

Create a new `IpMetadataChooser` that takes in the constructor:
- the mm_reader for dbip (and maxmind).
- the asn_db, as_to_type_map and as_to_org_map from Caida

This way you don't need all the different classes and interfaces. You are not gaining much from them. Just pass fake mm_readers and dicts in the tests.

Reply:
I think combining the different sub-pieces of the metadata classes here would be a mistake. 

Currently we have
file | lines
----|---
caida_ip_metadata.py | 370
test_caida_ip_metadata.py | 91
dbip.py | 70
maxmind.py | 116
mmdb_reader.py | 27
ip_metadata_chooser.py | 110

And in the future we plan to add more sub-metadata modules like AS size and possibly other whois information.

I think separating out
- how we get individual ip metadata from a specific source
from
- how we combine different ip metadata sources into a single addition to the row
is very important, and will only get more important in the future as we add more sources.

Reply:
A lot of that is boilerplate. You wouldn't lose any injection power.
But I'll leave that up to you. I do want, however, to have a factory function and the objects injected in the IPMetadataChooser constructor, since that's a lot clearer than having the two-stage initialization.

Reply:
Agreed about switching to the factory. I've added that.

I also figured out I can just subclass the fake off the real class instead of using an interface for both. That allows be to retain the type checking with less boilerplate.